Port the SOMA test group

xylar opened the pull request:
The Simulating Ocean Mesoscale Activity (SOMA) test group are a set of double gyre tests documented in [Wolfram et al. 2015](https://doi.org/10.1175/JPO-D-14-0260.1).

The `default` test cases use 100 vertical layers distributed according to layer thicknesses defined for some E3SMv1 meshes.

Other SOMA test cases (3-layer, time-varying wind and surface-restoring) will be ported as needed at a later time.

xylar commented:
## Testing

- [x] `soma/32km/default` on Ubuntu with gnu/mich
- [ ] `soma/16km/default` on Ubuntu with gnu/mich
- [ ] `soma/32km/default` on Chrysalis with intel/impi
- [ ] `soma/16km/default` on Chrysalis with intel/impi
- [ ] `soma/8km/default` on Chrysalis with intel/impi
- [ ] `soma/4km/default` on Chrysalis with intel/impi

xylar commented:
I need to add documentation for the SOMA test group.

xylar commented on compass/ocean/tests/soma/default/analysis.py:
```diff
@@ -0,0 +1,90 @@
+        s0 = section.getfloat('surface_salinity')
+
+        # hard coded factors from namelist file
+        gamma = 1.0 / 1250.0  # psu /m
+        beta = 0.05
+        zt = 300  # m
+        hshelf = 100  # m
+        h0 = 2400  # m
```
These would ideally not be hard-coded but instead would come from config options

pep8speaks commented:
Hello @xylar! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`compass/ocean/particles.py`](https://github.com/MPAS-Dev/compass/blob/e506dc5f7d80759d1bb12ccd30ceff8c19a7c714/compass/ocean/particles.py):

> [Line 185:36](https://github.com/MPAS-Dev/compass/blob/e506dc5f7d80759d1bb12ccd30ceff8c19a7c714/compass/ocean/particles.py#L185): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 188:40](https://github.com/MPAS-Dev/compass/blob/e506dc5f7d80759d1bb12ccd30ceff8c19a7c714/compass/ocean/particles.py#L188): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 193:52](https://github.com/MPAS-Dev/compass/blob/e506dc5f7d80759d1bb12ccd30ceff8c19a7c714/compass/ocean/particles.py#L193): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 194:52](https://github.com/MPAS-Dev/compass/blob/e506dc5f7d80759d1bb12ccd30ceff8c19a7c714/compass/ocean/particles.py#L194): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 195:52](https://github.com/MPAS-Dev/compass/blob/e506dc5f7d80759d1bb12ccd30ceff8c19a7c714/compass/ocean/particles.py#L195): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 196:53](https://github.com/MPAS-Dev/compass/blob/e506dc5f7d80759d1bb12ccd30ceff8c19a7c714/compass/ocean/particles.py#L196): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 197:59](https://github.com/MPAS-Dev/compass/blob/e506dc5f7d80759d1bb12ccd30ceff8c19a7c714/compass/ocean/particles.py#L197): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 202:61](https://github.com/MPAS-Dev/compass/blob/e506dc5f7d80759d1bb12ccd30ceff8c19a7c714/compass/ocean/particles.py#L202): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 203:57](https://github.com/MPAS-Dev/compass/blob/e506dc5f7d80759d1bb12ccd30ceff8c19a7c714/compass/ocean/particles.py#L203): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 204:61](https://github.com/MPAS-Dev/compass/blob/e506dc5f7d80759d1bb12ccd30ceff8c19a7c714/compass/ocean/particles.py#L204): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 205:59](https://github.com/MPAS-Dev/compass/blob/e506dc5f7d80759d1bb12ccd30ceff8c19a7c714/compass/ocean/particles.py#L205): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 206:52](https://github.com/MPAS-Dev/compass/blob/e506dc5f7d80759d1bb12ccd30ceff8c19a7c714/compass/ocean/particles.py#L206): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 217:35](https://github.com/MPAS-Dev/compass/blob/e506dc5f7d80759d1bb12ccd30ceff8c19a7c714/compass/ocean/particles.py#L217): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 219:35](https://github.com/MPAS-Dev/compass/blob/e506dc5f7d80759d1bb12ccd30ceff8c19a7c714/compass/ocean/particles.py#L219): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 221:35](https://github.com/MPAS-Dev/compass/blob/e506dc5f7d80759d1bb12ccd30ceff8c19a7c714/compass/ocean/particles.py#L221): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 223:42](https://github.com/MPAS-Dev/compass/blob/e506dc5f7d80759d1bb12ccd30ceff8c19a7c714/compass/ocean/particles.py#L223): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 227:36](https://github.com/MPAS-Dev/compass/blob/e506dc5f7d80759d1bb12ccd30ceff8c19a7c714/compass/ocean/particles.py#L227): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 230:57](https://github.com/MPAS-Dev/compass/blob/e506dc5f7d80759d1bb12ccd30ceff8c19a7c714/compass/ocean/particles.py#L230): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 239:48](https://github.com/MPAS-Dev/compass/blob/e506dc5f7d80759d1bb12ccd30ceff8c19a7c714/compass/ocean/particles.py#L239): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 241:38](https://github.com/MPAS-Dev/compass/blob/e506dc5f7d80759d1bb12ccd30ceff8c19a7c714/compass/ocean/particles.py#L241): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 243:37](https://github.com/MPAS-Dev/compass/blob/e506dc5f7d80759d1bb12ccd30ceff8c19a7c714/compass/ocean/particles.py#L243): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 245:74](https://github.com/MPAS-Dev/compass/blob/e506dc5f7d80759d1bb12ccd30ceff8c19a7c714/compass/ocean/particles.py#L245): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 246:74](https://github.com/MPAS-Dev/compass/blob/e506dc5f7d80759d1bb12ccd30ceff8c19a7c714/compass/ocean/particles.py#L246): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 247:74](https://github.com/MPAS-Dev/compass/blob/e506dc5f7d80759d1bb12ccd30ceff8c19a7c714/compass/ocean/particles.py#L247): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','





* In the file [`compass/ocean/tests/soma/default/analysis.py`](https://github.com/MPAS-Dev/compass/blob/e506dc5f7d80759d1bb12ccd30ceff8c19a7c714/compass/ocean/tests/soma/default/analysis.py):

> [Line 35:80](https://github.com/MPAS-Dev/compass/blob/e506dc5f7d80759d1bb12ccd30ceff8c19a7c714/compass/ocean/tests/soma/default/analysis.py#L35): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)

* In the file [`compass/ocean/tests/soma/forward.py`](https://github.com/MPAS-Dev/compass/blob/e506dc5f7d80759d1bb12ccd30ceff8c19a7c714/compass/ocean/tests/soma/forward.py):

> [Line 55:32](https://github.com/MPAS-Dev/compass/blob/e506dc5f7d80759d1bb12ccd30ceff8c19a7c714/compass/ocean/tests/soma/forward.py#L55): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 60:32](https://github.com/MPAS-Dev/compass/blob/e506dc5f7d80759d1bb12ccd30ceff8c19a7c714/compass/ocean/tests/soma/forward.py#L60): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 105:80](https://github.com/MPAS-Dev/compass/blob/e506dc5f7d80759d1bb12ccd30ceff8c19a7c714/compass/ocean/tests/soma/forward.py#L105): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)

xylar commented:
Note to self: make it easier to turn off particles.

xylar commented on compass/ocean/tests/soma/default/namelist.init:
```diff
@@ -0,0 +1,2 @@
+config_vert_levels = 100
+config_vertical_grid = '100layerE3SMv1'
```
```suggestion
config_vert_levels = 60
config_vertical_grid = '60layerPHC'
```
Maybe this is better?

xylar replied:
Maybe make this change in a follow-up PR?

xylar replied:
I think this should be
```suggestion
config_vert_levels = -1
config_vertical_grid = '60layerPHC'
config_soma_vert_levels = 60
```
I'll take a more careful look that the code to be sure.

ytakano3 replied:
@xylar Just to note (for myself) but as we talked, config_vert_levels = 60 did not work, did not change the information in the namelist so need to set config_coma_vert_levels = 60 as you suggested. This worked.

xylar commented on compass/ocean/tests/soma/namelist.init:
```diff
@@ -0,0 +1,5 @@
+config_ocean_run_mode = 'init'
+config_realistic_coriolis_parameter = .true.
+config_use_bulk_wind_stress = .true.
+config_realistic_coriolis_parameter = .true.
```
Redundant -- remove

xylar commented on compass/ocean/tests/soma/namelist.forward:
```diff
@@ -0,0 +1,27 @@
+config_use_cvmix_kpp = .true.
+config_use_bulk_wind_stress = .true.
+config_density0 = 1014.65
+config_eos_type = 'jm'
+config_eos_linear_alpha = 0.255
+config_eos_linear_beta = 0.764
+config_eos_linear_Tref = 19.0
+config_eos_linear_Sref = 35.0
```
It's odd that we're using JM but setting parameters for the linear EOS.  Also, the linear `alpha` here doesn't match the one used in the IC.

xylar commented on compass/ocean/tests/soma/forward.py:
```diff
@@ -0,0 +1,122 @@
+        """
+        self.resolution = resolution
+        self.with_particles = with_particles
+        res_params = {'32km': {'cores': 4,
+                               'min_cores': 1,
+                               'dt': "'00:10:00'",
+                               'btr_dt': "'0000_00:00:25'",
+                               'mom_del4': "2.0e11"},
+                      '16km': {'cores': 4,
+                               'min_cores': 1,
+                               'dt': "'00:10:00'",
+                               'btr_dt': "'0000_00:00:25'",
+                               'mom_del4': "2.0e10 "},
+                      '8km': {'cores': 4,
+                              'min_cores': 1,
+                              'dt': "'00:10:00'",
+                              'btr_dt': "'0000_00:00:15'",
+                              'mom_del4': "2.0e9"},
+                      '4km': {'cores': 4,
+                              'min_cores': 1,
+                              'dt': "'00:06:00'",
+                              'btr_dt': "'0000_00:00:10'",
+                              'mom_del4': "4.0e8"}}
```
We need to modify these.  Based on ~300 cells for the target number of cores and ~3000 for the minimum, we would have:
```suggestion
        res_params = {'32km': {'cores': 25,
                               'min_cores': 3,
                               'dt': "'00:10:00'",
                               'btr_dt': "'0000_00:00:25'",
                               'mom_del4': "2.0e11"},
                      '16km': {'cores': 100,
                               'min_cores': 10,
                               'dt': "'00:10:00'",
                               'btr_dt': "'0000_00:00:25'",
                               'mom_del4': "2.0e10 "},
                      '8km': {'cores': 400,
                              'min_cores': 40,
                              'dt': "'00:10:00'",
                              'btr_dt': "'0000_00:00:15'",
                              'mom_del4': "2.0e9"},
                      '4km': {'cores': 1600,
                              'min_cores': 160,
                              'dt': "'00:06:00'",
                              'btr_dt': "'0000_00:00:10'",
                              'mom_del4': "4.0e8"}}
```

xylar replied:
These should also be made config options instead of just hard coded.

xylar replied:
It would make sense to have `dt` be approximately proportional to the grid spacing.

xylar commented on compass/ocean/tests/soma/initial_state.py:
```diff
@@ -0,0 +1,106 @@
+        resolution : str
+            The resolution of the test case
+        """
+        super().__init__(test_case=test_case, name='initial_state')
```
This is implicitly defaulting to running on one core.  We may want to run on multiple cores, particularly at higher res.

xylar commented on compass/ocean/tests/soma/default/__init__.py:
```diff
@@ -0,0 +1,81 @@
+                                   out_name=out_name)
+        self.add_step(step)
+
+        with_particles = self.resolution == '32km'
```
This should be the default but we want a config option for turning particles on or off at all resolutions.

xylar replied:
It turns out to be difficult to make this a config option. Primarily, this is because the analysis step shouldn't happen if you're not using particles.  So I've made 2 versions of the test case at 32km resolution instead, one with and one without particles.  I think that's okay for now and we could add particles at other resolutions if we want to.

xylar commented on compass/ocean/tests/soma/namelist.forward:
```diff
@@ -0,0 +1,27 @@
+config_ocean_run_mode = 'forward'
+config_run_duration = '0000_00:01:30'
```
`dt` is at least 6 minutes, usually 10 minutes.  This should be updated to be several time steps.

mark-petersen commented:
@xylar I would like to preserve the original SOMA settings in the new repo in the history, and see you did that in 2f884182bf9f35ec1c697ce32efda113c2d13e97. As long as you keep that, you can proceed from there with additional commits, we can see the diffs, and you can add changes that make sense for your current setup in this PR. Since I have no personal experience with the settings, go ahead and change the settings above as you see fit.

xylar commented:
>  would like to preserve the original SOMA settings in the new repo in the history, and see you did that in 2f88418.
@mark-petersen, thanks for this comment.  In that commit "initial state" refers to the `initial_state` step, not the full test case.

I could squash some commits so that there is one that is a straight-up port of SOMA like you were expecting and then modify from there.  I think that's a good idea.

xylar commented:
I re-tested the 32km test case with and without particles after the many changes I just made and they both run fine.  The analysis plots look the same as before.

xylar commented on compass/ocean/tests/soma/streams.analysis:
```diff
@@ -0,0 +1,83 @@
+
+<stream name="lagrPartTrackRestart"/>
+
+<stream name="mixedLayerDepthsOutput"
+        precision="double"/>
```
We don't want to be writing this out, at least not every time step.

```suggestion
<stream name="mixedLayerDepthsOutput"
        precision="double"/
        output_interval="none"/>
```

xylar commented:
## Longer tests (2 years)

I'm also going to run all resolutions (without particles) for 2 years.
- [x] `soma/32km/default` on Chrysalis with intel/impi
- [ ] `soma/16km/default` on Chrysalis with intel/impi
- [ ] `soma/8km/default` on Chrysalis with intel/impi
- [ ] `soma/4km/default` on Chrysalis with intel/impi

I'll put relative vorticity plots here as they are done.

32 km:
![rel_vort_2yrs](https://user-images.githubusercontent.com/4179064/142651198-583d0e6a-a98e-4696-8f94-956b7284336d.png)

xylar commented:
@mark-petersen, while I'm running some longer tests, I think this is ready to review.

xylar commented:
@vanroekel, please feel free to review this, too, if you like and have time.  I won't add you but feel free to add yourself.

xylar commented:
@ytakano3, you had asked me how long you would need to run this simulation to reach quasi-equilibrium.  It seems like 2 years is not enough and you might need to run for 3 or 4.

You also asked me how long the simulation might take.  The answer so far (for 2 year runs) was about 1 hour 15 min for 32 km; 3 hours 30 minutes for 16 km.  I did a little optimization to reduce the output that might speed things up some.  I haven't finished the 8 or 4 km runs yet, but they're taking awhile even on a lot of cores.

xylar commented:
Here's a movie at 16 km so you can see what I mean about not quite being equilibrated:
https://drive.google.com/file/d/1cq5kA0cuTF8ck1D0Ktc3jBUfr9q6i7Ha/view?usp=sharing

xylar commented:
I rebased onto the latest `master` to get the updated E3SM-Project submodule (and other updates).

I also added a set of 3-year `long` test cases at each resolution.  I'm retesting with these runs now on Chrysalis (with intel and impi).

xylar commented:
@mark-petersen, I believe you haven't had a chance to test this PR yet.  I hope that's the case because I ported the remaining test cases today.  I was trying to see if this might be a good project for a student but it turned out not to be.  It required a lot of re-engineering for good code reuse.

There is now a single `SomaTestCase` class that can be used to create all SOMA test cases.  This class supports:
* different resolutions (4km, 8km, 16km and 32km)
* with or without particles
* 3 or 60 layers
* with or without surface restoring
* a short (3 time-step) or long (3 year) simulation time

For now, 4 variants are defined at most resolutions (`default`, `three_layer`, `surface_restoring` and `long`), with a 5th variant at 32km resolution that includes `particles`.  It would be very easy to add other variants if they turn out to be useful (e.g. `long_with_particles` or `three_layer_with_surface_restoring`) to SOMA's `__init__.py` module in the future.

It's up to you how you want to review this.  One option would be to set up the `legacy` versions of the test cases and compare the namelists and streams files with these test cases.  That's a little tricky because they are purposefully different in some cases (e.g. 60 layers instead of 100) but we could talk those through.

xylar commented:
@ytakano3, for your review, I think the best test would be if you could run the `ocean/soma/16km/long` test case for the full 3 years.  If you want to do that in one go, you can ask for 1 or 2 nodes on Chrysalis (or 2 or 3 on Anvil) for 24 hours.  If you want to run in shorter increments, you will need to change the following namelist options in order to restart the run:
```
&time_management
    config_do_restart = .true.
    ...
    config_start_time = 'file'
    ...
```

xylar commented:
Following up on https://github.com/MPAS-Dev/compass/pull/273#issuecomment-974444026, here is a movie at 4km resolution, showing that 3 years does seem to be long enough to reach quasi-equilibrium:
https://drive.google.com/file/d/1GHGwdefY9tJuSAcCiZAOB-cYmgL0Tf-A/view?usp=sharing

mark-petersen reviewed (commented):
@xylar thanks for all your work here. It is an impressive addition. I tested on cori with:
```
compass list|grep soma
 118: ocean/soma/4km/default
 119: ocean/soma/4km/long
 120: ocean/soma/4km/surface_restoring
 121: ocean/soma/4km/three_layer
 122: ocean/soma/8km/default
 123: ocean/soma/8km/long
 124: ocean/soma/8km/surface_restoring
 125: ocean/soma/8km/three_layer
 126: ocean/soma/16km/default
 127: ocean/soma/16km/long
 128: ocean/soma/16km/surface_restoring
 129: ocean/soma/16km/three_layer
 130: ocean/soma/32km/default
 131: ocean/soma/32km/long
 132: ocean/soma/32km/particles
 133: ocean/soma/32km/surface_restoring
 134: ocean/soma/32km/three_layer

compass setup -w $n/211206_test_soma_PR273 -n 130 131 132 133 134 -p $e/master/components/mpas-ocean/
```
I ran all the 32km cases on a single node on cori successfully, so I trust the mechanics work. I also looked through @xylar's comments above. Visually, things look right.

mark-petersen reviewed (approved):
I looked through the code, and I'm approving based on your testing, which looks great. I think if we find small flag settings to change in the next few months, it's best to just make another PR.  Thanks!

xylar commented:
I have rebased this branch and updated the documentation to include the 4 other variants of the test case besides `default`.  In the process of documenting `three_layer`, I realized that I missed a step in which 2 config options need to be set so that the continental shelf is removed.  I have fixed that and will re-test the 4 `three_layer` test cases to make sure they look right.

xylar commented:
Indeed, all `three_layer` test cases still run sucessfully on Chrysalis (with Intel and Intel-MPI).

The domain (`4km/three_layer` in blue) is smaller than for the other test cases (`4km/long` in white):
![three_layer_vs_long](https://user-images.githubusercontent.com/4179064/145559575-78e0d4a4-c691-4711-9edc-921f71f1b8ee.png)

xylar commented:
@mark-petersen, I assume you don't feel the need to revise your review based on these small changes?

mark-petersen commented:
no, please proceed. Thanks.

ytakano3 commented:
@xylar I finally managed to run SOMA 16km, long run and I will attach the relative vorticity figure from the year 3. I could not find the same color scale as yours but the range is the same. I think the structure looks similar and the simulation is working well (figure attached at the bottom). As you know, I had minor issue compiling but that is the character encoding setting on local machine and manage to change settings on cluster (where I ran SOMA). 

The other error is Low-level I/O error towards the end of the simulation but this does not impact the results as we communicate. SOMA 32km (short) test also looks good. Thank you!
![vtk_16km_year3](https://user-images.githubusercontent.com/47384280/145657798-090d192a-edba-4fa6-90dd-c4aa5506df0f.jpeg)

xylar commented:
Thanks @ytakano3! I appreciate you taking the time to run the model.

Just to document it, the errors @ytakano3 saw are:
```
PIO: ERROR: Changing the define mode for file (analysis_members/okuboWeiss.0003-01-01_00.00.00.nc) failed. Low-level I/O library API failed. NetCDF: Operation not allowed in data mode (error num=-38), (/gpfs/fs1/home/ac.xylar/mpas-work/compass/compass/conda/build_chrysalis_intel_impi/scorpio/src/clib/pioc_support.c:3377)
PIO: ERROR: Changing the define mode for file (analysis_members/highFrequencyOutput.0003-01-01.nc) failed. Low-level I/O library API failed. NetCDF: Operation not allowed in data mode (error num=-38), (/gpfs/fs1/home/ac.xylar/mpas-work/compass/compass/conda/build_chrysalis_intel_impi/scorpio/src/clib/pioc_support.c:3377)
PIO: ERROR: Changing the define mode for file (analysis_members/eliassenPalm.0003-01-01.nc) failed. Low-level I/O library API failed. NetCDF: Operation not allowed in data mode (error num=-38), (/gpfs/fs1/home/ac.xylar/mpas-work/compass/compass/conda/build_chrysalis_intel_impi/scorpio/src/clib/pioc_support.c:3377)
PIO: ERROR: Changing the define mode for file (analysis_members/mixedLayerDepths.0003-01-01.nc) failed. Low-level I/O library API failed. NetCDF: Operation not allowed in data mode (error num=-38), (/gpfs/fs1/home/ac.xylar/mpas-work/compass/compass/conda/build_chrysalis_intel_impi/scorpio/src/clib/pioc_support.c:3377)
PIO: ERROR: Changing the define mode for file (output/output.0003-01-01_00.00.00.nc) failed. Low-level I/O library API failed. NetCDF: Operation not allowed in data mode (error num=-38), (/gpfs/fs1/home/ac.xylar/mpas-work/compass/compass/conda/build_chrysalis_intel_impi/scorpio/src/clib/pioc_support.c:3377)
```
While I don't know exactly what the cause is, I have seen similar error messages and the output has still been okay.

@ytakano3, I don't know if you have permission, but the usual procedure to "officially" approve the PR would be to go to the "Files changes" tab at the top right, click on the green "Review changes", and enter your review.  That might be worth trying next time.
